### PR TITLE
Emit metrics to CloudWatch about the overall Agent and FileTailer health

### PR DESCRIPTION
**Pull request by diranged**

### CODE REVIEW

File: src/main/java/com/amazon/kinesis/streaming/agent/metrics/Metrics.java [hunk 1]
```diff
@@ -34,8 +34,9 @@
     public static final Pattern PARSER_TOTAL_RECORDS_PROCESSED_METRIC = Pattern.compile("^.*Parser.TotalRecordsProcessed$");
     public static final Pattern PARSER_TOTAL_RECORDS_SKIPPED_METRIC = Pattern.compile("^.*Parser.TotalRecordsSkipped$");
 
+    public static final String CLASS_DIMENSION = "Class";
     public static final String DESTINATION_DIMENSION = "Destination";
-    public static final String INSTANCE_DIMENSION = "Instance";
+    public static final String INSTANCE_DIMENSION = "InstanceId";
```
Comment (diranged) [inline src/main/java/com/amazon/kinesis/streaming/agent/metrics/Metrics.java:39]: I want to note here that the Amazon EC2 metrics all use the `InstanceId` dimension. In Datadog, we want (if possible) for our custom CloudWatch metrics to nicely tie directly to an EC2 Instance and all of its tags/etc.. so I am changing this. I need to talk to @Chaochenq about this though too.

**Comment by diranged** [comment]
@Chaochenq, What do you think about this btw? We've been running it for a few weeks now and it works great. The one thing I just added was the ability to whitelist different individual tailer metrics and agent metrics because we were sending way too many cloudwatch custom metrics (`InstanceId * MetricName * SourceFile`).

**Comment by chaochenq** [comment]
If I understand correctly, you are adding the 3 CW metrics to your dimension (per instance per stream)

FileTailer.FilesBehind
FileTailer.BytesBehind
FileTailer.RecordsTruncated

And the frequency of the emission is configured via "cloudwatch.metricGranularitySeconds". For example, if you configure this value to be 60 seconds, then the FileTailer.FilesBehind will be accumulated and emitted once every 60 seconds.

Does that sound correct?

**Comment by diranged** [comment]
@chaochenq yes - that is correct for the `FileTailer` metrics. In addition, I add a (defaulted `false`) configuration key to each Flow that allows you to selectively enable or disable these metrics. This is because we tail dozens of files on our hosts, and we were emitting thousands and thousands of custom metrics ($$$!)

Additionally we are passing the high level Agent metrics to CW. This is also disabled by default but something you can turn on. Also on this one we add a `metricWhitelist` so that you can select which metrics you want to send.

**Comment by chaochenq** [comment]
I see. I like the idea. Could you do the following?

1. instead of using "cloudwatch.emitMetrics" for controlling the emission of these health metrics, it's better to use a specific config key, like "cloudwatch.emitTailingMetrics". So that users can still turn on "cloudwatch.emitMetrics" to get the BytesSent metrics etc. but turn off "cloudwatch.emitTailingMetrics" to not emit the tailing metrics.

2. would it be better to have an enum for "cloudwatch.metricGranularitySeconds". I think only the values in the CW supported "period" value set makes sense. Right now CW allows you to select stat period like 1s, 5s, 10s, 30s, 1 min, 5 min, 15 min etc. If you configure this field to be 7 seconds, for example, then the Agent tailing CW metrics will be accumulated every 7 seconds. But if you look at your metrics in CW and select 10s or 1 min as period, the stat will be very hard to interpret. What do you think?

**Comment by diranged** [comment]
Let me think more about the second question. On the first one (the metric setting) - the reason I chose to put the `cloudwatch.emitMetrics` into each of the `Flow` configurations is so that you can decide which flows you might want metrics on vs others. In our case, we follow ~20 files ... but only 2 of the files we actually want detailed metrics on. For the other 18 files, we want to avoid having `18*3=54` custom metrics per host. Does that make sense?

**Comment by diranged** [comment]
@chaochenq here is a graph of our cloudwatch usage bill.. can you see when we turned on _full_ `FileTailer` and `Agent` metrics everywhere? :)

image

File: src/main/java/com/amazon/kinesis/streaming/agent/config/AgentConfiguration.java [hunk 2]
```diff
@@ -88,12 +93,21 @@ public boolean cloudwatchEmitMetrics() {
...
+    public List<String> cloudwatchMetricWhitelist() {
+        return this.readList("cloudwatch.metricWhitelist", String.class, DEFAULT_CW_METRIC_WHITELIST);
+    }
+
     public boolean cloudwatchTagInstance() {
         return this.readBoolean("cloudwatch.tagInstance",
                 DEFAULT_CW_TAG_INSTANCE);
     }
 
+    public int cloudwatchMetricGranularitySeconds() {
```
Comment (diranged) [inline src/main/java/com/amazon/kinesis/streaming/agent/config/AgentConfiguration.java:106]: @chaochenq,
  I agree with your comment - I'll make this an Enum tomorrow.

**Comment by chaochenq** [comment]
> Let me think more about the second question. On the first one (the metric setting) - the reason I chose to put the `cloudwatch.emitMetrics` into each of the `Flow` configurations is so that you can decide which flows you might want metrics on vs others. In our case, we follow ~20 files ... but only 2 of the files we actually want detailed metrics on. For the other 18 files, we want to avoid having `18*3=54` custom metrics per host. Does that make sense?

Yes I agree with putting this config key into each FileFlow. I was suggesting that you use a separate "cloudwatch.emitTailingMetrics" for each FileFlow and we still leave "cloudwatch.emitMetrics" in the general section so that this behavior doesn't change for users who are not aware of this new feature. Also we probably can rename "cloudwatch.metricWhitelist" to "cloudwatch.tailingMetrics". So the new config will look like this:

```
{
  "cloudwatch.emitMetrics": "true",
  "cloudwatch.tailingMetrics": ["TotalBytesBehind"],
  "fileFlow": {
    "filePattern": "logA",
    "cloudwatch.emitTailingMetrics": "true"
 }
```

With this config, customer will get CW metrics like

"BytesSent", "RecordsSent", "TotalBytesBehind", "FileTailer.FilesBehind" (for file "logA") etc.

```
{
  "cloudwatch.emitMetrics": "false",
  "cloudwatch.tailingMetrics": ["TotalBytesBehind"],
  "fileFlow": {
    "filePattern": "logA",
    "cloudwatch.emitTailingMetrics": "true"
 }
```

With this config above, customer won't get any CW metrics at all by turning CW emitMetrics off.

```
{
  "cloudwatch.emitMetrics": "true",
  "fileFlow": {
    "filePattern": "logA",
    "cloudwatch.emitTailingMetrics": "true"
 }
```

With this config above, customer will get CW metrics like

"BytesSent", "RecordsSent", "FileTailer.FilesBehind" (for file "logA") etc.

```
{
  "cloudwatch.emitMetrics": "true",
  "fileFlow": {
    "filePattern": "logA"
 }
```

With this config above, customer will get CW metrics like 

"BytesSent", "RecordsSent"

This is the same behavior as of today.

**Comment by diranged** [comment]
I'll dig into more of that next week I think. Right now we're running this release on all of our servers now and we are monitoring to see how the metric cost/usage drops.

**Pull request closed without merging**